File: frontend/src/api/explainer.ts
import { api } from "@/api/client";
import type { FlowEdge, FlowNode } from "@/types";

export type ExplainMode = "simple" | "line_by_line" | "interview" | "study_notes" | "generate_quiz" | "dry_run" | "custom";

export interface ExplainResponse {
  explanation: string;
  pseudocode: string;
  warnings: string[];
}

export async function explainFlowchart(
  nodes: FlowNode[],
  edges: FlowEdge[],
  mode: ExplainMode,
  customPrompt?: string
): Promise<ExplainResponse> {
  const { data } = await api.post<ExplainResponse>("/explainer/explain", {
    nodes: nodes.map((n) => ({ id: n.id, type: n.type, text: n.text })),
    edges: edges.map((e) => ({ id: e.id, fromNodeId: e.fromNodeId, toNodeId: e.toNodeId, label: e.label ?? null })),
    mode,
    custom_prompt: customPrompt,
  });
  return data;
}

File: frontend/api/app/api/routes/explainer.py
from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_current_user
from app.explainer.pipeline import explain_flowchart
from app.explainer.providers import TextProviderError, get_configured_text_provider
from app.explainer.schemas import ExplainRequest, ExplainResponse
from app.models.user import User

router = APIRouter(prefix="/explainer", tags=["explainer"])


@router.post("/explain", response_model=ExplainResponse)
def explain(payload: ExplainRequest, _: User = Depends(get_current_user)):
    try:
        provider = get_configured_text_provider()
    except TextProviderError as e:
        raise HTTPException(status_code=400, detail=str(e))

    try:
        result = explain_flowchart(
            [n.model_dump() for n in payload.nodes],
            [e.model_dump() for e in payload.edges],
            payload.mode,
            provider,
            payload.custom_prompt,
        )
    except Exception as e:
        err_str = str(e)
        if "quota" in err_str.lower() or "429" in err_str:
            raise HTTPException(status_code=429, detail="Gemini AI API rate limit reached (Quota Exceeded). Please wait a minute and try again!")
        raise HTTPException(status_code=502, detail=f"The AI explainer failed: {e}")

    return ExplainResponse(explanation=result.explanation, pseudocode=result.pseudocode, warnings=result.warnings)
